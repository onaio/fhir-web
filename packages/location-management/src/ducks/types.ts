import { Geometry } from 'geojson';
import { ParsedHierarchyNode, TreeNode } from './locationHierarchy/types';

/** The shape of a jurisdiction received from the OpenSRP API */
export interface OpenSRPJurisdiction {
  id: string;
  geometry?: Geometry;
  properties: {
    code?: string;
    geographicLevel: number;
    name: string;
    parentId?: string;
    status: string;
    version: string | number;
  };
  serverVersion: number;
  type: 'Feature';
}

export interface LocationTree {
  keys: React.Key[];
  node: ParsedHierarchyNode;
}

export type LocationTreeState = LocationTree;

/** Object containing known API endpoints by name */
export interface APIEndpoints {
  [key: string]: string;
}

/** Convenient type for either an array of jurisdictions or of tree nodes */
export type TreeNodeType = OpenSRPJurisdiction | TreeNode;
