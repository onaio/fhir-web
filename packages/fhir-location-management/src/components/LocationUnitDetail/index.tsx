import { Button, Alert, Spin, Col } from 'antd';
import React, { MouseEventHandler } from 'react';
import { get } from 'lodash';
import { CloseOutlined } from '@ant-design/icons';
import lang from '../../lang';
import { FHIRServiceClass, SingleKeyNestedValue } from '@opensrp/react-utils';
import { useQuery } from 'react-query';
import { ILocation } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/ILocation';
import { locationResourceType } from '../../constants';

export interface LUDProps {
  onClose: MouseEventHandler<HTMLElement>;
  fhirBaseUrl: string;
  detailId: string;
}

export const LocationUnitDetail: React.FC<LUDProps> = (props: LUDProps) => {
  const { onClose, fhirBaseUrl, detailId } = props;

  // fetch location details for the the selected location
  const serve = new FHIRServiceClass<ILocation>(fhirBaseUrl, locationResourceType);
  const { data, isLoading, error } = useQuery(
    [locationResourceType, detailId],
    async () => serve.read(detailId as string),
    {
      select: (res) => res,
      enabled: !!detailId,
    }
  );

  if (isLoading) {
    return <Spin  size='large' className="custom-spinner" />;
  }

  if (error && !data) {
    return <Alert data-testid="error-alert" message={`${error}`} type="error" />;
  }

  if (!data) {
    return <Alert data-testid="info-alert" message="Location resource not found" type="info" />;
  }

  const { name, status, description } = data;
  const detailsMap = {
    [lang.NAME]: name,
    [lang.STATUS]: status,
    [lang.DESCRIPTION]: description,
    [lang.PHYSICAL_TYPE]: get(data, 'physicalType.coding.0.display'),
    [lang.ALIAS]: get(data, 'alias.0'),
  };

  return (
    <Col className="pl-3" span={5}>
      <div className="p-4 bg-white" data-testid="view-details">
        <Button
          shape="circle"
          onClick={onClose}
          className="float-right"
          type="text"
          icon={<CloseOutlined />}
        />

        {Object.entries(detailsMap).map(([key, value]) => {
          const singleMapValue = { [key]: value };
          return (
            <div data-testid="single-key-value" key={key} className="pb-3">
              <SingleKeyNestedValue {...singleMapValue} />
            </div>
          );
        })}
      </div>
    </Col>
  );
};
