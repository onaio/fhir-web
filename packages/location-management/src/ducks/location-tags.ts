import {
  fetchActionCreatorFactory,
  removeActionCreatorFactory,
  setTotalRecordsFactory,
  reducerFactory,
  getItemsByIdFactory,
  getItemsArrayFactory,
  getItemByIdFactory,
  getTotalRecordsFactory,
} from '@opensrp/reducer-factory';

/** interface for location group **/
export interface LocationTag {
  id: number;
  active: boolean;
  name: string;
  description: string;
}

/** reducer name for the Item module */
export const reducerName = 'location-tags';

/** Action types */
const customfetchedActionType = 'location-tags/LOCATION_TAGS_FETCHED';
const customRemoveActionType = 'location-tags/REMOVE_LOCATION_TAGS';
const customSetTotalRecordsActionType = 'location-tags/SET_TOTAL_LOCATION_TAGS';

/** Item Reducer */
const reducer = reducerFactory<LocationTag>(
  reducerName,
  customfetchedActionType,
  customRemoveActionType,
  customSetTotalRecordsActionType
);

// action
/** actionCreator returns action to to add Item records to store */
export const fetchLocationTags = fetchActionCreatorFactory<LocationTag>(reducerName, 'id');
export const removeLocationTags = removeActionCreatorFactory(reducerName);
export const setTotalLocationtags = setTotalRecordsFactory(reducerName);

// selectors
export const getLocationTagsById = getItemsByIdFactory<LocationTag>(reducerName);
export const getLocationTagById = getItemByIdFactory<LocationTag>(reducerName);
export const getLocationTagsArray = getItemsArrayFactory<LocationTag>(reducerName);
export const getTotalLocationTags = getTotalRecordsFactory(reducerName);

export default reducer;
