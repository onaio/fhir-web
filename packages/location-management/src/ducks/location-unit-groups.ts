import {
  fetchActionCreatorFactory,
  removeActionCreatorFactory,
  setTotalRecordsFactory,
  reducerFactory,
  getItemsByIdFactory,
  getItemsArrayFactory,
  getItemByIdFactory,
  getTotalRecordsFactory,
} from '@opensrp/reducer-factory';

/** interface for location Unit Group **/
export interface LocationUnitGroup {
  id: number;
  active: boolean;
  name: string;
  description: string;
}

export interface LocationUnitGroupPayloadPOST {
  active: boolean;
  name: string;
  description: string;
}

export interface LocationUnitGroupPayloadPUT extends LocationUnitGroupPayloadPOST {
  id: string;
}

/** reducer name for the Item module */
export const reducerName = 'location-unit-groups';

/** Action types */
const customfetchedActionType = 'location-unit-groups/LOCATION_UNIT_GROUPS_FETCHED';
const customRemoveActionType = 'location-unit-groups/REMOVE_LOCATION_UNIT_GROUPS';
const customSetTotalRecordsActionType = 'location-unit-groups/SET_TOTAL_LOCATION_UNIT_GROUPS';

/** Item Reducer */
export const reducer = reducerFactory<LocationUnitGroup>(
  reducerName,
  customfetchedActionType,
  customRemoveActionType,
  customSetTotalRecordsActionType
);

// action
/** actionCreator returns action to to add Item records to store */
export const fetchLocationUnitGroups = fetchActionCreatorFactory<LocationUnitGroup>(
  reducerName,
  'id'
);
export const removeLocationUnitGroups = removeActionCreatorFactory(reducerName);
export const setTotalLocationunitgroups = setTotalRecordsFactory(reducerName);

// selectors
export const getLocationUnitGroupsById = getItemsByIdFactory<LocationUnitGroup>(reducerName);
export const getLocationUnitGroupById = getItemByIdFactory<LocationUnitGroup>(reducerName);
export const getLocationUnitGroupsArray = getItemsArrayFactory<LocationUnitGroup>(reducerName);
export const getTotalLocationUnitGroups = getTotalRecordsFactory(reducerName);

export default reducer;
