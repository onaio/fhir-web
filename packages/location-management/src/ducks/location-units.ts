import {
  fetchActionCreatorFactory,
  removeActionCreatorFactory,
  setTotalRecordsFactory,
  reducerFactory,
  getItemsByIdFactory,
  getItemsArrayFactory,
  getItemByIdFactory,
  getTotalRecordsFactory,
} from '@opensrp/reducer-factory';
import { Geometry } from 'geojson';
import { values } from 'lodash';
import { Store } from 'redux';

/** Enum representing the possible location unit status types */
export enum LocationUnitStatus {
  ACTIVE = 'Active',
  INACTIVE = 'InActive',
}

export enum LocationUnitSyncStatus {
  SYNCED = 'Synced',
  NOTSYNCED = 'NotSynced',
}

/** interface for LocationUnit.properties */
export interface Properties {
  name: string;
  parentId: string;
  status: LocationUnitStatus;
  geographicLevel?: number;
  username?: string;
  version?: number;
  name_en?: string;
  externalId?: string;
}

/** location unit tag interface */
export interface LocationUnitTag {
  id: number;
  name: string;
}

/** location interface */
export interface LocationUnit {
  id: string | number;
  properties: Properties;
  syncStatus: LocationUnitSyncStatus;
  type: string;
  locationTags?: LocationUnitTag[];
  geometry?: Geometry;
}

/** interface for the POST payload */
export interface LocationUnitPayloadPOST {
  properties?: Properties;
  syncStatus?: LocationUnitSyncStatus;
  type: string;
  locationTags?: LocationUnitTag[];
  geometry?: Geometry;
  textEntry?: string[];
  // we will remove this id as it should be auto generated on server
  id?: string | number;
}

/** interface for the PUT payload */
export interface LocationUnitPayloadPUT extends LocationUnitPayloadPOST {
  id: string | number;
}

/** reducer name for the Item module */
export const reducerName = 'location-units';

/** Action types */
const customfetchedActionType = 'location-units/LOCATION_UNITS_FETCHED';
const customRemoveActionType = 'location-units/REMOVE_LOCATION_UNITS';
const customSetTotalRecordsActionType = 'location-units/SET_TOTAL_LOCATION_UNITS';

/** Item Reducer */
const reducer = reducerFactory<LocationUnit>(
  reducerName,
  customfetchedActionType,
  customRemoveActionType,
  customSetTotalRecordsActionType
);

// action
/** actionCreator returns action to to add Item records to store */
export const fetchLocationUnits = fetchActionCreatorFactory<LocationUnit>(reducerName, 'id');
export const removeLocationUnits = removeActionCreatorFactory(reducerName);
export const setTotalLocationUnits = setTotalRecordsFactory(reducerName);

// selectors
export const getLocationUnitsById = getItemsByIdFactory<LocationUnit>(reducerName);
export const getLocationUnitById = getItemByIdFactory<LocationUnit>(reducerName);
export const getLocationUnitsArray = getItemsArrayFactory<LocationUnit>(reducerName);
export const getTotalLocationUnits = getTotalRecordsFactory(reducerName);

export const LocationUnitsArray = (state: Partial<Store>): LocationUnit[] =>
  values(getLocationUnitsById(state) || {});

export default reducer;
