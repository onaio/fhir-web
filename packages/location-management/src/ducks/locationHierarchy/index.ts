/** store location hierarchy information
 * - should be able to add a tree model
 * - modify the tree model in a manner such that shallow comparison would yield false
 * - provide the selectors
 */

import { Dictionary } from '@onaio/utils';
import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';
import { RawOpenSRPHierarchy, TreeNode } from './types';
import { generateJurisdictionTree } from './utils';
import { createSelector } from 'reselect';
import { values } from 'lodash';

/** reducer name for hierarchy reducer */
export const hierarchyReducerName = 'locationHierarchy';

// **************************** actions *****************************

/** action to add a tree to store */
export const TREE_FETCHED = 'opensrp/locations/hierarchy/TREE_FETCHED';

/** remove trees */
export const DEFOREST = 'opensrp/locations/hierarchy/DEFOREST';

/** describes action that adds a tree to store */
export interface FetchedTreeAction extends AnyAction {
  type: typeof TREE_FETCHED;
  treeByRootId: Dictionary<TreeNode>;
}

/** describes action to remove all trees */
export interface DeforestAction extends AnyAction {
  type: typeof DEFOREST;
  treeByRootId: {};
}

/** combined full action types | its a union */
export type TreeActionTypes = FetchedTreeAction | DeforestAction | AnyAction;

// **************************** action creators ****************************

/** action creator when adding a tree to store
 *
 * @param apiResponse - the raw hierarchy as received from opensrp
 * @param treeId - the treeId to use while saving to the store
 */
export function fetchTree(
  apiResponse: RawOpenSRPHierarchy,
  treeId: string | null = null
): FetchedTreeAction {
  const tree = generateJurisdictionTree(apiResponse);
  return {
    treeByRootId: {
      [treeId ? treeId : tree.model.id]: tree,
    },
    type: TREE_FETCHED,
  };
}

/** clear all the trees */
export function deforest(): DeforestAction {
  return {
    treeByRootId: {},
    type: DEFOREST,
  };
}

// **************************** reducer ****************************

/** The store's slice state
 */
export interface TreeState {
  treeByRootId: Dictionary<TreeNode> | {};
}

/** Create an immutable tree state */
export type ImmutableTreeState = TreeState & SeamlessImmutable.ImmutableObject<TreeState>;

/** starting state */
export const initialState: ImmutableTreeState = SeamlessImmutable({
  treeByRootId: {},
});

// the reducer function
/**
 *
 * @param {Dictionary} state - the store
 * @param {AnyAction} action - the redux action
 * @returns {object} - updated state
 */
export function hierarchyReducer(
  state: ImmutableTreeState | Dictionary = initialState,
  action: TreeActionTypes
) {
  switch (action.type) {
    case TREE_FETCHED:
      return {
        ...state,
        treeByRootId: { ...state.treeByRootId, ...action.treeByRootId },
      };
    case DEFOREST:
      return {
        ...state,
        treeByRootId: action.treeByRootId,
      };
    default:
      return state;
  }
}

/** prop filters to customize selector queries */
export interface Filters {
  rootJurisdictionId?: string[] /** specify which tree to act upon, undefined allows all */;
  geoLevel?: number /** get locations at the specified level */;
  searchQuery?: string /** search query to filter nodes against, by name and id */;
}

/** retrieve the rootJurisdiction value
 *
 * @param _ - the store
 * @param props -  the filterProps
 */
export const getRootJurisdictionId = (_: Partial<Store>, props: Filters) =>
  props.rootJurisdictionId;

/** retrieve the geoLevel value
 *
 * @param _ - the store
 * @param props -  the filterProps
 */
export const getGeoLevel = (_: Partial<Store>, props: Filters) => props.geoLevel;

/** retrieve the geoLevel value
 *
 * @param _ - the store
 * @param props -  the filterProps
 */
export const getSearchQuery = (_: Partial<Store>, props: Filters) => props.searchQuery;

/** gets all trees key'd by the rootNodes id
 *
 * @param state - the store
 * @param _ -  the filterProps
 */
export const getTreesByRootId = (state: Partial<Store>, _: Filters): Dictionary<TreeNode> =>
  (state as Dictionary)[hierarchyReducerName].treeByRootId;

/** factory that returns a selector to retrieve the tree(s) using their rootNode's ids */
export const getTreesByIds = () =>
  createSelector(getTreesByRootId, getRootJurisdictionId, (treesById, rootIds): TreeNode[] => {
    const allTreesArray = values(treesById);
    if (rootIds === undefined) {
      return allTreesArray;
    }
    const treesOfInterest: TreeNode[] = [];
    rootIds.forEach((rootId) => {
      const treeOfInterest = treesById[rootId] as TreeNode | undefined;
      if (treeOfInterest) {
        treesOfInterest.push(treeOfInterest);
      }
    });
    return treesOfInterest;
  });

const treeSelectors = getTreesByIds();

/** factory that returns a selector to retrieve the tree(s) using their rootNode's ids
 * the selector will return all nodes if geographic level is undefined
 */
export const getLocationsByLevel = () =>
  createSelector(treeSelectors, getGeoLevel, (trees, geoLevel): TreeNode[] => {
    let locationsOfInterest: TreeNode[] = [];
    trees.forEach((tree) => {
      locationsOfInterest = [
        ...locationsOfInterest,
        ...tree.all((node) => {
          if (geoLevel === undefined) {
            return true;
          }
          return node.model.node.attributes.geographicLevel === geoLevel;
        }),
      ];
    });
    return locationsOfInterest;
  });

const geoLevelSelector = getLocationsByLevel();

/** factory that returns a selector that can be used to filter the nodes by either their
 * labels or ids
 */
export const getLocationsByNameAndId = () =>
  createSelector(geoLevelSelector, getSearchQuery, (nodes, searchQuery): TreeNode[] => {
    if (searchQuery === undefined) {
      return nodes;
    }
    const matchesName = (node: TreeNode) =>
      node.model.label.toLowerCase().includes(searchQuery.toLowerCase());
    const matchesId = (node: TreeNode) => node.model.id === searchQuery;
    return nodes.filter((node) => {
      return matchesName(node) || matchesId(node);
    });
  });
