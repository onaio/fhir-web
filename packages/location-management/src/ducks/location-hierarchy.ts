/** store location hierarchy information
 * - should be able to add a tree model
 * - modify the tree model in a manner such that shallow comparison would yield false
 * - provide the selectors
 */

import { Dictionary } from '@onaio/utils';
import { AnyAction, Store } from 'redux';
import SeamlessImmutable from 'seamless-immutable';
import { TreeNode } from './locationHierarchy/types';
import { LocationTreeState } from './types';

/** reducer name for hierarchy reducer */
export const reducerName = 'location-hierarchy';

// **************************** actions *****************************

/** action to add a tree to store */
export const TREE_FETCHED = 'location-hierarchy/TREE_FETCHED';
export const FETCH_SINGLE_LOCATION = 'location-hierarchy/FETCH_SINGLE_LOCATION';
export const SET_LOCATION_TREE_STATE = 'location-hierarchy/SET_LOCATION_TREE_STATE';

/** describes action that adds a hierarchy tree to store */
export interface FetchedTreeAction extends AnyAction {
  type: typeof TREE_FETCHED;
  hierarchyObject: TreeNode;
}

/** describes action that saves a hierarchy tree to store */
export interface SetLocationTreeStateAction extends AnyAction {
  type: typeof SET_LOCATION_TREE_STATE;
  hierarchyObject: LocationTreeState;
}

/** combined full action types | its a union */
export type TreeActionTypes = FetchedTreeAction | SetLocationTreeStateAction | AnyAction;

// **************************** action creators ****************************

/** action creator when adding a tree to store
 *
 * @param {TreeNode} hierarchy - the raw hierarchy as received from opensrp
 * @returns {object} - action object
 */
export function fetchAllHierarchies(hierarchy: TreeNode): FetchedTreeAction {
  return {
    hierarchyObject: {
      ...hierarchy,
    } as TreeNode,
    type: TREE_FETCHED,
  };
}

/** action creator when adding a tree to store
 *
 * @param {LocationTreeState} hierarchy - the tree state to be expanded
 * @returns {object} - action object
 */
export function setLocationTreeState(hierarchy: LocationTreeState): SetLocationTreeStateAction {
  return {
    hierarchyObject: {
      ...hierarchy,
    } as LocationTreeState,
    type: SET_LOCATION_TREE_STATE,
  };
}

// **************************** medusa ****************************

/** The store's slice state */
export interface TreeState {
  hierarchyArray: TreeNode[];
  currentParentChildren: TreeNode[];
  locationTreeState: LocationTreeState;
}

/** Create an immutable tree state */
export type ImmutableTreeState = TreeState & SeamlessImmutable.ImmutableObject<TreeState>;

/** starting state */
export const initialState: ImmutableTreeState | Dictionary = SeamlessImmutable({
  hierarchyArray: [],
  currentParentChildren: [],
  locationObject: {},
  locationTreeState: {},
});

// the reducer function
/**
 *
 * @param {Dictionary} state - the store
 * @param {AnyAction} action - the redux action
 * @returns {object} - updated state
 */
export function reducer(
  state: ImmutableTreeState | Dictionary = initialState,
  action: TreeActionTypes
) {
  switch (action.type) {
    case TREE_FETCHED:
      return {
        ...state,
        hierarchyArray: [...state.hierarchyArray, action.hierarchyObject],
      };

    case SET_LOCATION_TREE_STATE:
      return {
        ...state,
        locationTreeState: action.hierarchyObject,
      };

    default:
      return state;
  }
}

/** gets array of all hierarchies
 *
 * @param {Store} state - the store
 * @returns {object} - returns item from location-hierarchy reducer
 */
export const getAllHierarchiesArray = (state: Partial<Store>): Dictionary<TreeNode> =>
  (state as Dictionary)[reducerName].hierarchyArray;

export const getLocationTreeState = (state: Partial<Store>): Dictionary<LocationTreeState> =>
  (state as Dictionary)[reducerName].locationTreeState;
