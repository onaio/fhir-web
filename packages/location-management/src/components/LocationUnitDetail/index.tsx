import { Button } from 'antd';
import React from 'react';
import { CloseOutlined } from '@ant-design/icons';
import { LocationUnit } from 'location-management/src/ducks/location-units';

export interface Props extends LocationUnit {
  onClose?: Function;
}

const LocationUnitDetail: React.FC<Props> = (props: Props) => {
  return (
    <div className="p-4 bg-white">
      <Button
        shape="circle"
        onClick={() => (props.onClose ? props.onClose() : '')}
        className="float-right"
        type="text"
        icon={<CloseOutlined />}
      />
      <div className="mb-4 small mt-4">
        <p className="mb-0 font-weight-bold">Name</p>
        <p className="mb-0 loc-desc">{props.properties.name}</p>
      </div>

      <div className="mb-4 small">
        <p className="mb-0 font-weight-bold">Status</p>
        <p className="mb-0 loc-desc">{props.properties.status}</p>
      </div>

      <div className="mb-4 small">
        <p className="mb-0 font-weight-bold">Type</p>
        <p className="mb-0 loc-desc">{props.type}</p>
      </div>

      <div className="mb-4 small">
        <p className="mb-0 font-weight-bold">External Id</p>
        <p className="mb-0 loc-desc">{props.properties.externalId}</p>
      </div>

      <div className="mb-4 small">
        <p className="mb-0 font-weight-bold">Username</p>
        <p className="mb-0 loc-desc">{props.properties.username}</p>
      </div>

      <div className="mb-4 small">
        <p className="mb-0 font-weight-bold">Version</p>
        <p className="mb-0 loc-desc">{props.properties.version}</p>
      </div>

      <div className="mb-4 small">
        <p className="mb-0 font-weight-bold">Sync status</p>
        <p className="mb-0 loc-desc">{props.syncStatus}</p>
      </div>

      <div className="mb-4 small">
        <p className="mb-0 font-weight-bold">Level</p>
        <p className="mb-0 loc-desc">{props.properties.geographicLevel}</p>
      </div>
    </div>
  );
};

export default LocationUnitDetail;
