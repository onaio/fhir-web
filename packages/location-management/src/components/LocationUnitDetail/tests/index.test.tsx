import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import LocationUnitDetail, { Props } from '..';
import { LocationUnitStatus, LocationUnitSyncStatus } from '../../../ducks/location-units';

describe('location-management/src/components/LocationUnitDetail', () => {
  const props: Props = {
    id: '6y66ggde',
    type: 'Feautire',
    properties: {
      parentId: '2',
      name: 'test',
      status: LocationUnitStatus.ACTIVE,
      externalId: 'asdkjh1230',
      username: 'edward 0',
      version: 0,
      geographicLevel: 2,
    },
    syncStatus: LocationUnitSyncStatus.SYNCED,
  };

  it('renders without crashing', () => {
    const wrapper = shallow(<LocationUnitDetail {...props} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('removes it self on close', () => {
    const wrapper = mount(<LocationUnitDetail {...props} onClose={() => wrapper.unmount()} />);
    expect(wrapper.children()).toHaveLength(1);
    wrapper.find('button').simulate('click');

    wrapper.update();
    expect(wrapper).toHaveLength(0);
  });

  it('doesnt remove itself if onclose is not provided', () => {
    const wrapper = mount(<LocationUnitDetail {...props} />);
    expect(wrapper.children()).toHaveLength(1);
    wrapper.find('button').simulate('click');
    expect(wrapper).toHaveLength(1);
  });
});
