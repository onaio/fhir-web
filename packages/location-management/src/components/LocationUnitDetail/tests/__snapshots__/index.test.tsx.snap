// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/pages/Home renders without crashing 1`] = `
<LocationUnitDetail
  id="6y66ggde"
  properties={
    Object {
      "externalId": "asdkjh1230",
      "geographicLevel": 2,
      "name": "test",
      "parentId": "2",
      "status": "Active",
      "username": "edward 0",
      "version": 0,
    }
  }
  syncStatus="Synced"
  type="Feautire"
>
  <div
    className="p-4 bg-white"
  >
    <Button
      block={false}
      className="float-right"
      ghost={false}
      htmlType="button"
      icon={<ForwardRef(CloseOutlined) />}
      loading={false}
      onClick={[Function]}
      shape="circle"
      type="text"
    >
      <button
        className="ant-btn ant-btn-text ant-btn-circle ant-btn-icon-only float-right"
        onClick={[Function]}
        type="button"
      >
        <ForwardRef(CloseOutlined)>
          <AntdIcon
            icon={
              Object {
                "icon": Object {
                  "attrs": Object {
                    "focusable": "false",
                    "viewBox": "64 64 896 896",
                  },
                  "children": Array [
                    Object {
                      "attrs": Object {
                        "d": "M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z",
                      },
                      "tag": "path",
                    },
                  ],
                  "tag": "svg",
                },
                "name": "close",
                "theme": "outlined",
              }
            }
          >
            <span
              aria-label="close"
              className="anticon anticon-close"
              role="img"
            >
              <IconReact
                className=""
                icon={
                  Object {
                    "icon": Object {
                      "attrs": Object {
                        "focusable": "false",
                        "viewBox": "64 64 896 896",
                      },
                      "children": Array [
                        Object {
                          "attrs": Object {
                            "d": "M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z",
                          },
                          "tag": "path",
                        },
                      ],
                      "tag": "svg",
                    },
                    "name": "close",
                    "theme": "outlined",
                  }
                }
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="close"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  key="svg-close"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                    key="svg-close-svg-0"
                  />
                </svg>
              </IconReact>
            </span>
          </AntdIcon>
        </ForwardRef(CloseOutlined)>
      </button>
    </Button>
    <div
      className="mb-4 small mt-4"
    >
      <p
        className="mb-0 font-weight-bold"
      >
        Name
      </p>
      <p
        className="mb-0 loc-desc"
      >
        test
      </p>
    </div>
    <div
      className="mb-4 small"
    >
      <p
        className="mb-0 font-weight-bold"
      >
        Status
      </p>
      <p
        className="mb-0 loc-desc"
      >
        Active
      </p>
    </div>
    <div
      className="mb-4 small"
    >
      <p
        className="mb-0 font-weight-bold"
      >
        Type
      </p>
      <p
        className="mb-0 loc-desc"
      >
        Feautire
      </p>
    </div>
    <div
      className="mb-4 small"
    >
      <p
        className="mb-0 font-weight-bold"
      >
        External Id
      </p>
      <p
        className="mb-0 loc-desc"
      >
        asdkjh1230
      </p>
    </div>
    <div
      className="mb-4 small"
    >
      <p
        className="mb-0 font-weight-bold"
      >
        Username
      </p>
      <p
        className="mb-0 loc-desc"
      >
        edward 0
      </p>
    </div>
    <div
      className="mb-4 small"
    >
      <p
        className="mb-0 font-weight-bold"
      >
        Version
      </p>
      <p
        className="mb-0 loc-desc"
      >
        0
      </p>
    </div>
    <div
      className="mb-4 small"
    >
      <p
        className="mb-0 font-weight-bold"
      >
        Sync status
      </p>
      <p
        className="mb-0 loc-desc"
      >
        Synced
      </p>
    </div>
    <div
      className="mb-4 small"
    >
      <p
        className="mb-0 font-weight-bold"
      >
        Level
      </p>
      <p
        className="mb-0 loc-desc"
      >
        2
      </p>
    </div>
  </div>
</LocationUnitDetail>
`;
